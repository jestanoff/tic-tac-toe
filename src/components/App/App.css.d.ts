// This file is automatically generated.
// Please do not change this file!
interface CssExports {
  'bluePrimary': string;
  'board': string;
  'container': string;
  'flip': string;
  'fontPrimary': string;
  'greySecondary': string;
  'greyTertiary': string;
  'innerContainer': string;
  'version': string;
}
export const cssExports: CssExports;
export default cssExports;
