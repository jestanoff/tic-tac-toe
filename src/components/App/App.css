@value bluePrimary, greySecondary from "css/colors.css";

.container {
  background: greySecondary;
  box-shadow: 0 0 2px rgba(0, 0, 0, 0.4);
  margin: 50px auto;
  max-width: 500px;
  min-width: 240px;
  position: relative;
}

.board {
  background-color: bluePrimary;
  left: 0;
  position: relative;
  transition: transform 0.5s;
}

.flip {
  transform: rotateY(-180deg);
}

.innerContainer {
  margin: 0 auto;
  padding: 16px 0;
  position: relative;
  width: 240px;
}

:global {
  .example-enter {
    top: -500px;
  }

  .example-enter.example-enter-active {
    top: 0px;
    transition: top 0.3s ease-out;
  }
}
