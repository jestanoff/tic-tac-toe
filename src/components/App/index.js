import React, { memo, useCallback, useEffect, useMemo, useRef, useState } from 'react';
import { CSSTransition, TransitionGroup } from 'react-transition-group';
import Button from '../Button';
import Board from '../Board';
import Header from '../Header';
import NotificationBar from '../NotificationBar';
import WinningLine from '../WinningLine';
import ScoresSection from '../ScoreSection';
import Select from '../Select';
import Fireworks from '../Fireworks';
import { isGameOver, computeAIMove, getNotification } from '../../helpers';
import {
  AI_WAITING_TIME,
  DRAW,
  PLAYER_O,
  PLAYER_X,
  NUM_OF_CELLS,
  SYMBOLS,
  UNRESOLVED,
  EASY,
  HARD,
  RESET,
  DARK_GRAY,
  WHITE,
} from '../../constants';
import styles from './App.css';
import '../../css/font.css';

const initialState = {
  boardStatus: Array(NUM_OF_CELLS).fill(0),
  difficulty: EASY,
  history: [{ board: Array(NUM_OF_CELLS).fill(0) }],
  isBoardUiDisabled: false,
  notification: 'Start game by clicking on any cell',
  outcome: { winner: UNRESOLVED, line: UNRESOLVED },
  playerTurn: PLAYER_X,
};
const options = [EASY, HARD];

const TicTacToeApp = () => {
  const [
    { boardStatus, difficulty, isBoardUiDisabled, notification, outcome, playerTurn, timeoutID },
    setState,
  ] = useState(initialState);
  const board = useRef(null);

  useEffect(() => {
    if (outcome.winner === UNRESOLVED && playerTurn === PLAYER_O) {
      const id = setTimeout(() => {
        const AIBoardStatus = computeAIMove(boardStatus, difficulty);
        const AIOutcome = isGameOver(AIBoardStatus);

        setState((prevState) => ({
          ...prevState,
          boardStatus: AIBoardStatus,
          isBoardUiDisabled: false,
          notification: getNotification(AIOutcome.winner),
          playerTurn: PLAYER_X,
          outcome: AIOutcome,
          history: [...history, { board: AIBoardStatus }],
        }));
      }, AI_WAITING_TIME);

      setState((prevState) => ({ ...prevState, timeoutID: id }));
    }
  }, [boardStatus, difficulty, outcome, playerTurn]);

  const resetGame = useCallback(() => {
    if (board.current) {
      const cx = board.current.classList;
      cx.contains(styles.flip) ? cx.remove(styles.flip) : cx.add(styles.flip);
      clearTimeout(timeoutID);
      setState((prevState) => ({ ...initialState, difficulty: prevState.difficulty }));
    }
  }, [timeoutID]);

  const handleCellClick = useCallback(
    (cellIndex) => {
      if (boardStatus[cellIndex] === 0 && !isBoardUiDisabled) {
        const nextBoardStatus = boardStatus.slice();
        nextBoardStatus[cellIndex] = PLAYER_X;
        const nextOutcome = isGameOver(nextBoardStatus);
        setState((prevState) => ({
          ...prevState,
          boardStatus: nextBoardStatus,
          isBoardUiDisabled: true,
          notification: getNotification(nextOutcome.winner),
          playerTurn: PLAYER_O,
          outcome: nextOutcome,
          history: [...history, { board: nextBoardStatus }],
        }));
      }

      if (outcome.winner === DRAW) resetGame();
    },
    [boardStatus, isBoardUiDisabled, outcome, resetGame],
  );

  const handleDifficultyChange = useCallback(
    (event) => {
      console.log('event.target.value', event.target.value);
      setState((prevState) => ({
        ...prevState,
        difficulty: event.target.value,
      }));
      resetGame();
    },
    [resetGame],
  );

  const showIcon = boardStatus.some(Boolean) && outcome.winner !== 0;
  const icon = SYMBOLS[outcome.winner] || SYMBOLS[playerTurn];
  const showWinningLine = outcome.line > UNRESOLVED;
  const timeout = useMemo(() => ({ enter: 300, exit: 1 }), []);

  return (
    <TransitionGroup>
      <CSSTransition classNames="example" timeout={timeout}>
        <div className={styles.container} id="main-container">
          {outcome.winner === PLAYER_X && <Fireworks handleClick={resetGame} />}
          <Select current={difficulty} options={options} onChange={handleDifficultyChange} />
          <Header title="Tic Tac Toe" />
          <ScoresSection outcome={outcome.winner} playerTurn={playerTurn} />
          <NotificationBar icon={icon} msg={notification} showIcon={showIcon} />
          <section className={styles.board} ref={board}>
            <div className={styles.innerContainer}>
              {showWinningLine && (
                <WinningLine
                  line={outcome.line}
                  handleClick={resetGame}
                  color={outcome.winner === PLAYER_X ? DARK_GRAY : WHITE}
                />
              )}
              <Board boardStatus={boardStatus} handleCellClick={handleCellClick} />
            </div>
          </section>
          <Button text={RESET} handleClick={resetGame} />
        </div>
      </CSSTransition>
    </TransitionGroup>
  );
};

export default memo(TicTacToeApp);
