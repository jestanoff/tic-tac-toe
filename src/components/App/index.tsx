import * as React from 'react';
import { CSSTransition, TransitionGroup } from 'react-transition-group';
import Button from 'components/Button';
import Board from 'components/Board';
import Header from 'components/Header';
import NotificationBar from 'components/NotificationBar';
import WinningLine from 'components/WinningLine';
import ScoreBoard from 'components/ScoreBoard';
import Select from 'components/Select';
import Fireworks from 'components/Fireworks';
import { isGameOver, computeAIMove, getNotification } from 'helpers';
import {
  AI_WAITING_TIME,
  DRAW,
  PLAYER_O,
  PLAYER_X,
  NUM_OF_CELLS,
  SYMBOLS,
  UNRESOLVED,
  EASY,
  HARD,
  RESET,
  DARK_GRAY,
  WHITE,
} from '../../constants';
import 'css/font.css';
import styles from './App.css';

interface Board {
  board: Array<number>;
}

interface Outcome {
  winner: number;
  line: number;
}

interface State {
  boardStatus: Array<number>;
  difficulty: string;
  history: Array<Board>;
  isBoardUiDisabled: boolean;
  notification: string;
  outcome: Outcome;
  playerTurn: number;
  timeoutId?: ReturnType<typeof setTimeout>;
}

const initialState: State = {
  boardStatus: Array(NUM_OF_CELLS).fill(0),
  difficulty: EASY,
  history: [{ board: Array(NUM_OF_CELLS).fill(0) }],
  isBoardUiDisabled: false,
  notification: 'Start game by clicking on any cell',
  outcome: { winner: UNRESOLVED, line: UNRESOLVED },
  playerTurn: PLAYER_X,
};
const options = [EASY, HARD];

const TicTacToeApp = () => {
  const [
    { boardStatus, difficulty, isBoardUiDisabled, notification, outcome, playerTurn, timeoutId },
    setState,
  ] = React.useState(initialState);
  const board = React.useRef<HTMLElement>(null);

  React.useEffect(() => {
    if (outcome.winner === UNRESOLVED && playerTurn === PLAYER_O) {
      const id = setTimeout(() => {
        const AIBoardStatus = computeAIMove(boardStatus, difficulty);
        const AIOutcome = isGameOver(AIBoardStatus);

        setState(
          (prevState: State): State => ({
            ...prevState,
            boardStatus: AIBoardStatus,
            isBoardUiDisabled: false,
            notification: getNotification(AIOutcome.winner),
            playerTurn: PLAYER_X,
            outcome: AIOutcome,
            history: [...prevState.history, { board: AIBoardStatus }],
          }),
        );
      }, AI_WAITING_TIME);

      setState((prevState) => {
        return { ...prevState, timeoutId: id };
      });
    }
  }, [boardStatus, difficulty, outcome, playerTurn]);

  const resetGame = React.useCallback(() => {
    if (board.current) {
      const cx = board.current?.classList ?? '';
      cx.contains(styles.flip) ? cx.remove(styles.flip) : cx.add(styles.flip);
      setState((prevState) => ({ ...initialState, difficulty: prevState.difficulty }));
    }

    if (timeoutId) {
      clearTimeout(timeoutId);
    }
  }, [timeoutId]);

  const handleSpaceClick = React.useCallback(
    (cellIndex) => {
      if (boardStatus[cellIndex] === 0 && !isBoardUiDisabled) {
        const nextBoardStatus = boardStatus.slice();
        nextBoardStatus[cellIndex] = PLAYER_X;
        const nextOutcome = isGameOver(nextBoardStatus);
        setState((prevState) => ({
          ...prevState,
          boardStatus: nextBoardStatus,
          isBoardUiDisabled: true,
          notification: getNotification(nextOutcome.winner),
          playerTurn: PLAYER_O,
          outcome: nextOutcome,
          history: [...prevState.history, { board: nextBoardStatus }],
        }));
      }

      if (outcome.winner === DRAW) resetGame();
    },
    [boardStatus, isBoardUiDisabled, outcome, resetGame],
  );

  const handleDifficultyChange = React.useCallback(
    (event) => {
      setState((prevState) => ({
        ...prevState,
        difficulty: event.target.value,
      }));

      resetGame();
    },
    [resetGame],
  );

  const showIcon = boardStatus.some(Boolean) && outcome.winner !== 0;
  const icon = SYMBOLS[outcome.winner] || SYMBOLS[playerTurn];
  const showWinningLine = outcome.line > UNRESOLVED;
  const timeout = React.useMemo(() => ({ enter: 300, exit: 1 }), []);

  return (
    <TransitionGroup>
      <CSSTransition classNames="example" timeout={timeout}>
        <div className={styles.container} id="main-container">
          {outcome.winner === PLAYER_X && <Fireworks handleClick={resetGame} />}
          <Select current={difficulty} options={options} onChange={handleDifficultyChange} />
          <Header title="Tic Tac Toe" />
          <ScoreBoard outcome={outcome.winner} playerTurn={playerTurn} />
          <NotificationBar icon={icon} msg={notification} showIcon={showIcon} />
          <section className={styles.board} ref={board}>
            <div className={styles.innerContainer}>
              {showWinningLine && (
                <WinningLine
                  line={outcome.line}
                  handleClick={resetGame}
                  color={outcome.winner === PLAYER_X ? DARK_GRAY : WHITE}
                />
              )}
              <Board boardStatus={boardStatus} handleSpaceClick={handleSpaceClick} />
            </div>
          </section>
          <Button text={RESET} handleClick={resetGame} />
        </div>
      </CSSTransition>
    </TransitionGroup>
  );
};

export default React.memo(TicTacToeApp);
