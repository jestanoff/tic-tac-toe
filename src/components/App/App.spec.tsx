import * as React from 'react';
import { render, screen } from '@testing-library/react'
import App from './index';

describe('<App />', () => {
  test('should render the initial state of the game board with difficulty selector and start notification', () => {
    const { asFragment } = render(<App />);

    expect(screen.getByRole('combobox', { name: /Difficulty selector/i })).toBeInTheDocument();
    expect(screen.getByRole('heading', { level: 1 })).toHaveTextContent('Tic Tac Toe');
    expect(screen.getByRole('heading', { level: 2 })).toHaveTextContent('Start the game by selecting any empty space');
    expect(screen.getByRole('button', { name: /^Top left empty space/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /^Top center empty space/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /^Top right empty space/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /^Left middle empty space/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /^Center middle empty space/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /^Right middle empty space/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /^Bottom left empty space/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /^Bottom center empty space/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /^Bottom right empty space/i })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: /Reset/i })).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  test('should show difficulty selector with "Easy" selected by default', async () => {
    expect.assertions(3);
    render(<App />);

    await screen.findByRole('combobox', { name: /Difficulty selector/i }); // Not actually async but just in case it becomes
    const easyOption: HTMLOptionElement = screen.getByRole('option', { name: /Easy/i });
    expect(easyOption).toBeInTheDocument();
    expect(easyOption.selected).toBe(true)
    expect(screen.getByRole('option', { name: /Impossible/i })).toBeInTheDocument();
  });

  test('should show the score board with initially no score for each player', () => {
    render(<App />);

    expect(screen.getByTestId('score-for-player-cross')).toHaveTextContent('—');
    expect(screen.getByTitle('Player cross has 0 points')).toBeInTheDocument();
    expect(screen.getByTestId('score-for-player-circle')).toHaveTextContent('—');
    expect(screen.getByTitle('Player circle has 0 points')).toBeInTheDocument();
  });

  test('should display a notification message helping the human player make its first move', () => {
    render(<App />);

    expect(screen.getByRole('heading', { level: 2 })).toHaveTextContent('Start the game by selecting any empty space');
    expect(screen.getByTitle('Player cross can start the game by selecting any empty space')).toBeInTheDocument();
  });
  
  test('should show "Reset" button that is disabled if no spaces has been marked', () => {
    render(<App />);

    expect(screen.getByRole('button', { name: /Reset/i })).toBeDisabled();
  });
});
