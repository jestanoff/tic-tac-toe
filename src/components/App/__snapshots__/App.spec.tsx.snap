// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> should render the initial state of the game board with difficulty selector and start notification 1`] = `
<DocumentFragment>
  <div>
    <div
      class="container"
      id="main-container"
    >
      <section
        class="container"
      >
        <select
          title="Difficulty selector"
        >
          <option
            value="Easy"
          >
            Easy
          </option>
          <option
            value="Impossible"
          >
            Impossible
          </option>
        </select>
      </section>
      <h1
        class="heading"
      >
        Tic Tac Toe
      </h1>
      <section
        class="container"
      >
        <div
          class="active"
          data-testid="score-for-player-cross"
        >
          <span
            class="mark"
          >
            <div
              class="container"
            >
              <svg
                class="svg"
                data-testid="cross-mark"
                preserveAspectRatio="xMinYMin meet"
                version="1.1"
                viewBox="0 0 500 500"
              >
                <line
                  stroke="#444444"
                  stroke-width="50"
                  x1="130"
                  x2="130"
                  y1="130"
                  y2="130"
                >
                  <animate
                    attributeName="x2"
                    attributeType="XML"
                    dur="1ms"
                    fill="freeze"
                    from="130"
                    id="first"
                    repeatCount="1"
                    to="370"
                  />
                  <animate
                    attributeName="y2"
                    attributeType="XML"
                    dur="1ms"
                    fill="freeze"
                    from="130"
                    repeatCount="1"
                    to="370"
                  />
                </line>
                <line
                  opacity="0"
                  stroke="#444444"
                  stroke-width="50"
                  x1="370"
                  x2="370"
                  y1="130"
                  y2="130"
                >
                  <animate
                    attributeName="opacity"
                    attributeType="XML"
                    begin="0s"
                    dur="1ms"
                    fill="freeze"
                    from="0"
                    repeatCount="1"
                    to="1"
                  />
                  <animate
                    attributeName="x2"
                    attributeType="XML"
                    begin="0s"
                    dur="1ms"
                    fill="freeze"
                    from="370"
                    repeatCount="1"
                    to="130"
                  />
                  <animate
                    attributeName="y2"
                    attributeType="XML"
                    begin="0s"
                    dur="1ms"
                    fill="freeze"
                    from="130"
                    repeatCount="1"
                    to="370"
                  />
                </line>
              </svg>
            </div>
          </span>
          <span
            aria-live="polite"
            class="score"
            title="Player cross has 0 points"
          >
            —
          </span>
        </div>
        <div
          class="container"
          data-testid="score-for-player-circle"
        >
          <span
            class="mark"
          >
            <div
              class="container"
            >
              <svg
                class="svg"
                data-testid="circle-mark"
                preserveAspectRatio="xMinYMin meet"
                version="1.1"
                viewBox="0 0 500 500"
              >
                <circle
                  cx="250"
                  cy="250"
                  fill="transparent"
                  r="125"
                  stroke="#444444"
                  stroke-dasharray="800"
                  stroke-dashoffset="-800"
                  stroke-width="45"
                >
                  <animate
                    attributeName="stroke-dashoffset"
                    attributeType="XML"
                    dur="1ms"
                    fill="freeze"
                    from="-800"
                    repeatCount="1"
                    to="0"
                  />
                </circle>
              </svg>
            </div>
          </span>
          <span
            aria-live="polite"
            class="score"
            title="Player circle has 0 points"
          >
            —
          </span>
        </div>
      </section>
      <h2
        class="container"
      >
        <span
          aria-live="polite"
          title="Player cross can start the game by selecting any empty space"
        >
          Start the game by selecting any empty space
        </span>
      </h2>
      <section
        class="board"
      >
        <div
          class="innerContainer"
        >
          <section
            class="container"
          >
            <button
              class="space top left"
              title="Top left empty space"
            />
            <button
              class="space top center"
              title="Top center empty space"
            />
            <button
              class="space top right"
              title="Top right empty space"
            />
            <button
              class="space left middle"
              title="Left middle empty space"
            />
            <button
              class="space center middle"
              title="Center middle empty space"
            />
            <button
              class="space right middle"
              title="Right middle empty space"
            />
            <button
              class="space bottom left"
              title="Bottom left empty space"
            />
            <button
              class="space bottom center"
              title="Bottom center empty space"
            />
            <button
              class="space bottom right"
              title="Bottom right empty space"
            />
          </section>
        </div>
      </section>
      <div
        class="container"
      >
        <button
          class="button"
          disabled=""
        >
          Reset
        </button>
      </div>
      <p
        class="version"
        title="App version"
      >
        v1.0.1
      </p>
    </div>
  </div>
</DocumentFragment>
`;
